### PR TITLE
feat(issuer-irec-api): import irec certificates

### PR DESCRIPTION
**Pull request by ioncreature**
- [x] return list of certificates to import
- [x] tests for list
- [x] import certificate
- [x] tests for import

### CODE REVIEW

File: packages/organizations/origin-organization-irec-api/migrations/1626682324674-UserName.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import { MigrationInterface, QueryRunner } from 'typeorm';
+
+export class UserName1626682324674 implements MigrationInterface {
+    name = 'UserName1626682324674';
+
+    public async up(queryRunner: QueryRunner): Promise<void> {
+        await queryRunner.query(
+            `ALTER TABLE "irec_connection" ADD "userName" character varying NOT NULL`
```
Comment (soanvig) [inline packages/organizations/origin-organization-irec-api/migrations/1626682324674-UserName.ts:8]: This will fail on real database because it's NOT NULL

File: packages/traceability/issuer-irec-api-wrapper/test/flow.test.ts [hunk 2]
```diff
@@ -107,7 +107,7 @@ describe('API flows', () => {
         expect(device.status).to.equal(DeviceState.Draft);
     }).timeout(15000);
 
-    it('should pass create and approve issue flow', async () => {
+    it.only('should pass create and approve issue flow', async () => {
```
Comment (soanvig) [inline packages/traceability/issuer-irec-api-wrapper/test/flow.test.ts]: got ya

**Review by soanvig — DISMISSED** [review]

File: packages/organizations/origin-organization-irec-api/migrations/1626682324674-UserName.ts (hunk 1 above)
Comment (josipbagaric) [reply packages/organizations/origin-organization-irec-api/migrations/1626682324674-UserName.ts:8]: Sure, but I guess it's not a big issue since it was never used on a real db? @ioncreature

File: packages/traceability/issuer-irec-api/src/pods/certificate/certificate.controller.ts [hunk 3]
```diff
@@ -39,4 +52,27 @@ export class IrecCertificateController extends CertificateController {
...
+    @UseGuards(AuthGuard(), ActiveUserGuard, RolesGuard, ActiveOrganizationGuard)
+    @Roles(Role.OrganizationAdmin, Role.OrganizationDeviceManager)
+    @ApiBody({ type: ImportIrecCertificateDTO })
+    @ApiResponse({
+        status: HttpStatus.CREATED,
+        type: DeviceDTO,
+        description: 'Imports a certificate from IREC'
+    })
+    @ApiForbiddenResponse({
+        status: HttpStatus.FORBIDDEN,
```
Comment (josipbagaric) [inline packages/traceability/issuer-irec-api/src/pods/certificate/certificate.controller.ts]: No need to define this, since ApiForbiddenResponse defaults to FORBIDDEN anyway.

File: packages/traceability/issuer-irec-api/src/pods/certificate/certificate.controller.ts [hunk 4]
```diff
@@ -39,4 +52,27 @@ export class IrecCertificateController extends CertificateController {
...
+        description: `User doesn't have the correct permissions`
+    })
+    @ApiBadRequestResponse({ status: HttpStatus.BAD_REQUEST })
```
Comment (josipbagaric) [inline packages/traceability/issuer-irec-api/src/pods/certificate/certificate.controller.ts]: Same here + missing some description on why it could happen.

File: packages/traceability/issuer-irec-api/src/pods/certificate/handler/import-irec-certificate.handler.ts [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        if (!originDevice) {
+            throw new BadRequestException('Unknown IREC device');
+        }
+
+        const platformAdmin = await this.userService.getPlatformAdmin();
+        const transaction = await this.irecService.transferCertificate(
+            user,
+            platformAdmin.organization.id,
+            certificateToImport.assetId
+        );
```
Comment (josipbagaric) [inline packages/traceability/issuer-irec-api/src/pods/certificate/handler/import-irec-certificate.handler.ts:66]: Is this where we check if the user owns the asset it's trying to import?

File: packages/organizations/origin-organization-irec-api/migrations/1626682324674-UserName.ts (hunk 1 above)
Comment (ioncreature) [reply packages/organizations/origin-organization-irec-api/migrations/1626682324674-UserName.ts:8]: Yep, it was never used.

File: packages/traceability/issuer-irec-api/src/pods/certificate/certificate.controller.ts (hunk 3 above)
Comment (ioncreature) [reply packages/traceability/issuer-irec-api/src/pods/certificate/certificate.controller.ts]: removed

File: packages/traceability/issuer-irec-api/src/pods/certificate/certificate.controller.ts (hunk 4 above)
Comment (ioncreature) [reply packages/traceability/issuer-irec-api/src/pods/certificate/certificate.controller.ts]: removed

File: packages/traceability/issuer-irec-api/src/pods/certificate/handler/import-irec-certificate.handler.ts (hunk 5 above)
Comment (ioncreature) [reply packages/traceability/issuer-irec-api/src/pods/certificate/handler/import-irec-certificate.handler.ts:66]: ownership check is on line 40

**Review by josipbagaric — APPROVED** [review]

File: packages/traceability/issuer-api/src/pods/certification-request/handlers/get-all-certification-requests.handler.ts [hunk 6]
```diff
@@ -16,6 +16,11 @@ export class GetAllCertificationRequestsHandler
     ) {}
 
     async execute({ query }: GetAllCertificationRequestsQuery): Promise<CertificationRequestDTO[]> {
-        return this.repository.find(query);
+        const q = query || {};
+        return this.repository.find({
+            ...(q.hasOwnProperty('approved') ? { approved: q.approved } : {}),
+            ...(q.owner ? { owner: q.owner } : {}),
+            ...(q.deviceIds?.length ? { deviceId: In(q.deviceIds) } : {})
+        });
```
Comment (josipbagaric) [inline packages/traceability/issuer-api/src/pods/certification-request/handlers/get-all-certification-requests.handler.ts:24]: What's the reason for this change? It now only allows these 3 properties to be queried, and not others.

File: packages/traceability/issuer-irec-api/test/certificate.e2e-spec.ts [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+        expect(certificateToImport.asset).to.equal('test-asset-id');
+
+        await request(app.getHttpServer())
+            .post(`/irec/certificate/import-certificate`)
+            .send({ assetId: certificateToImport.asset })
+            .set({ 'test-user': TestUser.OrganizationDeviceManager })
+            .expect(HttpStatus.CREATED);
+
+        await sleep(10000);
+    });
```
Comment (josipbagaric) [inline packages/traceability/issuer-irec-api/test/certificate.e2e-spec.ts:336]: Isn't this use case the only one that's I-REC related? Do we need all the tests above? They seem like duplicates to me to the ones in the issuer-api.

Do they also test some I-REC functionality?

Comment (ioncreature) [reply packages/traceability/issuer-irec-api/test/certificate.e2e-spec.ts:336]: I've added them to check if all non irec specific stuff works. And some of them failed at that time. So they are required because it helps me not to break non irec functionality

File: packages/traceability/issuer-api/src/pods/certification-request/handlers/get-all-certification-requests.handler.ts (hunk 6 above)
Comment (ioncreature) [reply packages/traceability/issuer-api/src/pods/certification-request/handlers/get-all-certification-requests.handler.ts:24]: We can add other fields later when we need it

File: packages/traceability/issuer-irec-api/test/certificate.e2e-spec.ts (hunk 7 above)
Comment (josipbagaric) [reply packages/traceability/issuer-irec-api/test/certificate.e2e-spec.ts:336]: Ok. Just not sure why we need these 2 tests:

1. `should create a private certificate`
2. `should transfer a private certificate`

Are private certificates even supported in I-REC?

Comment (ioncreature) [reply packages/traceability/issuer-irec-api/test/certificate.e2e-spec.ts:390]: Yep, there are no private/public certificates. Ok, ill remove tests for this, but what about the controller methods? Do I have to remove them too?

Comment (ioncreature) [reply packages/traceability/issuer-irec-api/test/certificate.e2e-spec.ts:336]: No, there is no such functionality. Ill remove those tests
